Game logic merge request

Mikearaya opened the pull request:
On this pull request, I have added all the features that make up the tic tac toe game such as :
- Creating a player
- assigning unique symbols based on each player  preference
- recording game statistics
- validating all user input's
- generating a game report
- exception handling
- colorize text to create a better user experience

### I Kindly request your review so I can merge to development branch ad start working on other parts of the program

**Thank you** :bowing_man:

Mikearaya committed:
Create Board class with main functionalities

Mikearaya committed:
Create player class

Mikearaya committed:
Create game class

Mikearaya committed:
Refactor game logic from main.rb

Mikearaya committed:
Create custom exception class

Mikearaya committed:
Update move validator

Mikearaya committed:
Add custom validation and user input validation

Mikearaya committed:
Add symbole validator

Mikearaya committed:
Merge with user_interface branch

Mikearaya committed:
Add colorization for text

Mikearaya committed:
Refactor game class

majovanilla commented on lib/custom_exception.rb:
```diff
@@ -0,0 +1,6 @@
+require 'colorize'
```
You are using this gem but is not on your Gemfile. Anyone who tries to run your game won't be able to do it since you didn't include it here. The app will through an error and it's not very user friendly.

Mikearaya replied:
**Fixed**

majovanilla commented on lib/player.rb:
```diff
@@ -0,0 +1,15 @@
+class Player
+  attr_reader :name
+  attr_reader :symbole
```
The correct word in English is `symbol`. Please correct this on the user interface when you ask the player to choose the symbol.

Mikearaya replied:
**Fixed**

majovanilla commented on lib/validatable.rb:
```diff
@@ -0,0 +1,16 @@
+
+  def self.valid_symbole?(symbole)
+    unless symbole.match(/^[\D]{1}$/)
+      raise CustomException, "Invalid Symbole \n Valid Symbole is single none digit characted\n Try Again"
```
The correct word is `character`. Also, check the other comment about `symbole`.

**Optional** I would advise you to not allow the players to change the symbol and stick with the traditional X and O to avoid unexpected behavior.

Mikearaya replied:
I have fixed the spelling error. and for symbols preference, I think there will be any unexpected behavior because there is only one way to set and read the value. who knows someone might have a lucky character :smile:

majovanilla commented on lib/tic_tac_toe.rb:
```diff
@@ -0,0 +1,134 @@
+  end
+
+  def add_players
+    print 'First player name: '.blue
+    first_player_name = gets.chomp
+    print ' Symbole: '.blue
+    symbole = read_symbole
+    @players << Player.new(first_player_name, symbole)
+    print 'Second player name: '.blue
+    second_player_name = gets.chomp
+    print ' Symbole: '.blue
+    symbole = read_symbole
+    @players << Player.new(second_player_name, symbole)
```
No puts/print and get.chomp outside main.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from Pathwright)

Please check your method. Right now your game allows a player to leave the name empty and this shouldn't happen.

Mikearaya replied:
**Fixed**, I thought it would be more organized this way.

majovanilla commented on lib/tic_tac_toe.rb:
```diff
@@ -0,0 +1,134 @@
+        symbole = gets.chomp
+
+        Validatable.valid_symbole?(symbole)
+        if @used_symbole[":#{symbole}"]
+          puts 'symbole already used, try different'
+          next
+        end
+        @used_symbole[":#{symbole}"] = true
+        break
+      rescue CustomException => e
+        puts e.display_error
+      rescue StandardError => e
+        puts e.message
```
No puts/print and get.chomp outside main.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from Pathwright)

Mikearaya replied:
**Fixed**

majovanilla commented on lib/tic_tac_toe.rb:
```diff
@@ -0,0 +1,134 @@
+
+  def change_turn
+    @turn = @turn == 2 ? 1 : 2
+    puts get_player(@turn).name + '\'s Turn'
```
No puts/print and get.chomp outside main.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from Pathwright)

Mikearaya replied:
**Fixed**

majovanilla commented on lib/tic_tac_toe.rb:
```diff
@@ -0,0 +1,134 @@
+    choice = 0
+    loop do
+      begin
+        choice = gets.chomp
+        raise CustomException, 'Invalid Input, Valid Move should be digit between 1-9' unless validate_move(choice.to_s)
+
+        choice = choice.to_i
+        break
+      rescue CustomException => e
+        puts "#{e.display_error}  \n Try Again"
+      rescue StandardError => e
+        puts "#{e.message}  \n Try Again"
+      end
+    end
```
No puts/print and get.chomp outside main.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from Pathwright)

Mikearaya replied:
**Fixed**

majovanilla commented on lib/tic_tac_toe.rb:
```diff
@@ -0,0 +1,134 @@
+  end
+
+  def display_winner_message(winner)
+    puts <<-HEARDOC
+        ************************************************************
+        *                    Congradulation                        *
+                      #{get_player(winner).name}  has won
+        ************************************************************
+    HEARDOC
```
No puts/print and get.chomp outside main.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from Pathwright)

I really liked this part, but it's not displaying correctly on the terminal (some elements appear aligned to the right and some to the left) You should improve this to make it more readable.

You should display the winning board before you print this message. Right now the players can't see it.

Mikearaya replied:
**Fixed** all I/O to be on main have added minor adjustment to text alignment and color, I'll go through what is left on the next milestone (Final touch up) :smile:

majovanilla commented on lib/tic_tac_toe.rb:
```diff
@@ -0,0 +1,134 @@
+  public
+
+  def display_statstics
+    puts <<-HEARDOC
+
+
+    ************************************************************
+                        GAME STATISTICS
+
+                        Total Matches #{@total_match}
+
+              #{@players[0].name}    #{@players[1].name}
+      Wins    #{@players[0].wins}           #{@players[1].wins}
+
+    ************************************************************
+    HEARDOC
```
No puts/print and get.chomp outside main.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from Pathwright)

Mikearaya replied:
**Fixed**

majovanilla commented on lib/board.rb:
```diff
@@ -0,0 +1,51 @@
+    @tile[index - 1] = symbole
+  end
+
+  def won?(symbole)
+    return false if @total_moves < 4
```
You are not considering the possibility when no player wins and the game ends in a draw. Right now, your game keeps asking the player to choose a cell but they are all empty.

There can't be a winner before move 5, not 4.

Mikearaya replied:
**Fixed** was counting from zero :wink:

majovanilla commented on bin/main.rb:
```diff
@@ -1,87 +1,34 @@
+  game.play
+  puts 'Game Over \n Do you want to play another round (y/n)'
+
+  loop do
+    begin
+    play = gets.chomp
+    Validatable.exit_choice_valid?(play)
+    continue = play.downcase != 'n'
```
Your loop's indentation is not correct. Please make it right so the code is more readable.

Mikearaya replied:
**Fixed**

majovanilla reviewed (changes requested):
Hi @Mikearaya 

## You did a *great* job on this project! 💜 

I liked your PR message!

There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- Remember the requirement about not having puts/print or get.chomp outside your main.rb file. I commented the code below, but I put it here in case I missed some lines.

- The previous doesn't mean that you can't return a string inside your class methods and then use it to print it in your main.rb :wink: 

### Optional Changes:

If you want to make your project even better 🔥💥, I left a few suggestions you can implement.  These are completely [Optional]: 🦔
- Your colored messages are a nice touch, but sometimes I found them too strong for me. Especially the one with a white background and red letters. You could use a much friendlier version for the eyes :smile: 

## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻

Mikearaya committed:
Refactored new player input value to be on main.rb

Mikearaya committed:
Refactored code to make prompts for input on main.rb

Mikearaya committed:
Fix spelling errors

Mikearaya committed:
Adjust text color and alignment

Mikearaya committed:
Edit text color

Mikearaya commented:
Thank you :bowing_man:  @majovanilla for your thorough review and recommendations.

I have fixed all of the issues on the previous review such as:
- moving all I/O to be handled inside `main.rb`. I thought it was modularized to let each class handle I/O related to it previously :smile: 
- detecting draw scenarios (no one wins)
- fixed loud text colors to a more eye-friendly

### Now I'm kindly requesting another review so I can merge to develop and start working on finalizing the project 

**Thank you** :bowing_man:

majovanilla commented:
Hi @Mikearaya  remember to request it on the dashboard! :smile:  

Good luck with the next review!

euqueme commented on lib/tic_tac_toe.rb:
```diff
@@ -0,0 +1,104 @@
+    false
+  end
+
+  def display_winner_message
+    puts <<-HEARDOC
+
+    ************************************************************
+    *                    Congratulations                        *
+                      #{get_player(@turn).name}  has won
+
+    HEARDOC
+    display_statstics
+  end
```
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- As it was suggested by @majovanilla you can return a String variable instead and print it in your main.rb file

Mikearaya replied:
**Fixed**, it must have slipped :smiley:

euqueme reviewed (approved):
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @Mikearaya :handshake: 

### Thank you for doing the changes requested by @majovanilla  :medal_sports:  this is one of the best tic tac toe games I've seen so far you should feel very proud of yourself  not only for the super friendly gameplay :muscle: :video_game: but also you've followed OOP rules to the teeth, you won't have problems testing your code in a future project so claps to you :clap: :clap: :clap: :1st_place_medal: 

There is still one minor change to make, please address it in the comments in your code, I won't stop you for this but please fix it before merging this branch

### That's it! :raised_hands: 

### Please :pray: keep up the good work :clinking_glasses: :tada: :tada: :tada: 

PROJECT STATUS: APPROVED :heavy_check_mark:

Mikearaya committed:
Refactor display_winner_message method

Mikearaya commented:
@euqueme first I want to Thank you for being reasonable and letting me fix the issue without failing the review.

**And Thank you for acknowledging my efforts, this is by far the best positive feedback I've received on my pull request :bow: :blush: :100: I really appreciate it**
